src/core/types/message.rs

```diff
@@ -4,17 +4,177 @@ use crate::core::types::{
-#[serde(rename_all = "snake_case")]
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
 #[serde(untagged)]
-pub enum MessageType {}
+pub enum MessageKind {
+    IncomingMessage {
+        #[serde(flatten)]
+        forward_kind: ForwardKind,
+        edit_date: Option<i32>,
+        #[serde(flatten)]
+        media_kind: MediaKind,
+        reply_markup: Option<InlineKeyboardMarkup>,
+    },
+    NewChatMembers {
+        new_chat_members: Vec<User>,
+    },
+    LeftChatMember {
+        left_chat_member: User,
+    },
+    NewChatTitle {
+        new_chat_title: String,
+    },
+    NewChatPhoto {
+        new_chat_photo: Vec<PhotoSize>,
+    },
+    DeleteChatPhoto {
+        delete_chat_photo: bool,
+    },
+    GroupChatCreated {
+        group_chat_created: bool,
+    },
+    SupergroupChatCreated {
+        supergroup_chat_created: bool,
+    },
+    ChannelChatCreated {
+        channel_chat_created: bool,
+    },
+    Migrate {
+        migrate_to_chat_id: i64,
+        migrate_from_chat_id: i64,
+    },
+    PinnedMessage {
+        pinned_message: Box<Message>,
+    },
+    Invoice {
+        invoice: Invoice,
+    },
+    SuccessfulPayment {
+        successful_payment: SuccessfulPayment,
+    },
+    ConnectedWebsite {
+        connected_website: String,
+    },
+    PassportData {
+        passport_data: PassportData,
+    },
+}
+
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(untagged)]
+pub enum ForwardKind {
+    ChannelForward {
+        forward_from_channel: Box<Chat>,
+        forward_from_message_id: i64,
+        forward_signature: String,
+    },
+    NonChannelForward {
+        #[serde(flatten)]
+        forward: UserOrSenderName,
+    },
+    Origin {
+        reply_to_message: Option<Box<Message>>,
+    },
+}
+
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(untagged)]
+pub enum UserOrSenderName {
+    User { from: User },
+    SenderName { sender_name: String },
+}
+
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(untagged)]
+pub enum MediaKind {
+    Animation {
+        animation: Animation,
+        caption: Option<String>,
+    },
+    Audio {
+        audio: Audio,
+        caption: Option<String>,
+    },
+    Contact {
+        contact: Contact,
+    },
+    Document {
+        document: Document,
+        caption: Option<String>,
+    },
+    Game {
+        game: Game,
+    },
+    Location {
+        location: Location,
+    },
+    Photo {
+        sizes: Vec<PhotoSize>,
+        caption: Option<String>,
+    },
+    Poll {
+        poll: Poll,
+    },
+    Sticker {
+        sticker: Sticker,
+    },
+    Text {
+        text: String,
+        entities: Vec<MessageEntity>,
+    },
+    Video {
+        video: Video,
+        caption: Option<String>,
+    },
+    VideoNote {
+        video_note: VideoNote,
+    },
+    Voice {
+        voice: Voice,
+        caption: Option<String>,
+    },
+    Venue {
+        venue: Venue,
+    },
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
`Message` structure is big and also there are a lot of changes from telegram's structure, so we need more tests I think. You can use `SendMessage` in this crate and telegram bot [@ShowJsonBot](https://t.me/@ShowJsonBot) for json examples

Reply:
I'll make more tests later.